lib/util/files/find.js

```diff
@@ -0,0 +1,72 @@
+"use strict";
+
+let fs = require("fs");
+let path = require("path");
+
+let stat = promisify(fs.stat);
+let readdir = promisify(fs.readdir);
+
+// determines files within a directory, optionally filtered by a function
+//
+// if the returned function is called without any arguments, it will return a
+// Promise with an array of all files that matched
+// if it is called with an array of paths, it will return a Promise with an
+// array of all provided paths that match
+module.exports = function findFiles(sourceDir, filter = (_ => true)) {
+	return files => {
+		return determineFilesToProcess(sourceDir, files).
+			then(skipDotfiles).
+			then(fileNames => fileNames.filter(filter));
+	};
+};
+
+function determineFilesToProcess(directory, files) {
+	return files ? filesWithinDirectory(directory, files) : tree(directory);
+}
+
+function skipDotfiles(files) {
+	return files.filter(filename => !filename.startsWith("."));
+}
+
+function filesWithinDirectory(directory, files) {
+	return new Promise(resolve => {
+		resolve(files.
+			map(filepath => path.relative(directory, filepath)).
+			filter(filename => !filename.startsWith("..")));
+	});
+}
+
+function tree(filepath, referenceDir = filepath) {
+	return stat(filepath).
+		then(res => {
+			if(!res.isDirectory()) {
+				return [ path.relative(referenceDir, filepath) ];
+			}
+
+			return readdir(filepath).
+				then(entries => {
+					let res = Promise.all(entries.map(entry => {
+						return tree(path.join(filepath, entry), referenceDir);
+					}));
+					return res.then(flatten);
+				});
+		});
+}
+
+function promisify(fn) {
```

Comment:
Should we maybe make this available as a helper module? I would use it in -sass, -files and -images.

Reply:
I suppose that makes sense (93fa2837aad0a1c7839e7dc88e529c4548767e0c's hesitation is effectively superseded by d783fc4f3e6ad3393e3c96acc4051dc88adeb29e).

There's probably a gazillion dedicated npm packages for this, but there's value in remaining self-contained for such ~~simple~~small stuff.